File: iteration#1/src/system/University.java
package system;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import dto.CourseDto;
import dto.PrerequisiteDto;
import dto.StudentDto;
import model.Advisor;
import model.Course;
import model.Student;
import model.StudentTranscript;

import java.io.FileWriter;
import java.io.IOException;
import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

import static java.util.List.of;

public class University {
    private ArrayList<Student> students;
    private ArrayList<Course> courses;
    private ArrayList<Course> nteCourses;
    private ArrayList<Course> fteCourses;
    private ArrayList<Course> teCourses;
    private ArrayList<Advisor> advisors;
    private List<String> studentNames;
    private ArrayList<StudentTranscript> studentTranscripts;
    private String semester;
    private Boolean randomPlay;


    public University() {
        this.students = new ArrayList<Student>();
        this.courses = new ArrayList<Course>();
        this.fteCourses = new ArrayList<Course>();
        this.nteCourses = new ArrayList<Course>();
        this.teCourses = new ArrayList<Course>();
        this.advisors = new ArrayList<Advisor>();
        this.studentTranscripts = new ArrayList<StudentTranscript>();
        semester = "fall";
        //it will change once we read the input.json.
        studentNames = new ArrayList<String>();
        randomPlay = true;
    }

    public void registerStudent(Student student){
        students.add(student);
    }

    public void hireAdvisor(Advisor advisor) {
        advisors.add(advisor);
    }

    public void addNewCourse(Course course){
        this.courses.add(course);
    }

    public void readAllInput() throws IOException {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        Reader reader = Files.newBufferedReader(Paths.get("iteration#1/data/input.json"));

        Map<?,?> map = gson.fromJson(reader,Map.class);

        List<CourseDto> courseDtos = new ArrayList<CourseDto>();
        List<CourseDto> teCourseDtos = new ArrayList<CourseDto>();
        List<CourseDto> nteCourseDtos = new ArrayList<CourseDto>();
        List<CourseDto> fteCourseDtos = new ArrayList<CourseDto>();

        List<PrerequisiteDto> prerequisiteDtos = new ArrayList<PrerequisiteDto>();

        String batchProcessPlay = "true";

        for(Map.Entry<?,?> entry:map.entrySet()){
            String key = (String) entry.getKey();
            if(key.equals("semester")){//
                semester = (String) entry.getValue();
            }
            else if(key.equals("nonTechnicalElectiveCourses")){
                JsonArray jsonArray = gson.toJsonTree(entry.getValue()).getAsJsonArray();
                nteCourseDtos = Arrays.asList(gson.fromJson(jsonArray,CourseDto[].class));
            }
            else if(key.equals("compulsoryCourses")){
                JsonArray jsonArray = gson.toJsonTree(entry.getValue()).getAsJsonArray();
                courseDtos = Arrays.asList(gson.fromJson(jsonArray,CourseDto[].class));
            }
            else if(key.equals("technicalElectiveCourses")){
                JsonArray jsonArray = gson.toJsonTree(entry.getValue()).getAsJsonArray();
                teCourseDtos = Arrays.asList(gson.fromJson(jsonArray,CourseDto[].class));
            }
            else if(key.equals("facultyTechnicalElectiveCourses")){
                JsonArray jsonArray = gson.toJsonTree(entry.getValue()).getAsJsonArray();
                fteCourseDtos = Arrays.asList(gson.fromJson(jsonArray,CourseDto[].class));
            }
            else if(key.equals("prerequisites")){//
                JsonArray jsonArray = gson.toJsonTree(entry.getValue()).getAsJsonArray();
                prerequisiteDtos = Arrays.asList(gson.fromJson(jsonArray,PrerequisiteDto[].class));
            }
            else if(key.equals("studentNames")){//
                JsonArray jsonArray = gson.toJsonTree(entry.getValue()).getAsJsonArray();
                studentNames = Arrays.asList(gson.fromJson(jsonArray,String[].class));
            }
            else if(key.equals("batchProcess")){//
                batchProcessPlay = (String)entry.getValue();
            }
        }
        this.randomPlay = (batchProcessPlay.equals("true")) ? true : false;
        reader.close();
        String[] temp = {
                "Prof. Dr. Haluk Rahmi TOCUOGLU",
                "Prof.Dr. Cigdem Eroglu ERDEM",
                "Doc. Dr. Ali Fuat ALKAYA",
                "Doc. Dr. Murat Can GANIZ",
                "Doc. Dr. Mujdat SOYTURK",
                "Doc. Dr. Borahan TUMER",
                "Doc. Dr. Mustafa AGAOGLU",
                "Dr. Ogr. Uyesi Mehmet BARAN",
                "Dr. Ogr. Uyesi Betul Demiroz BOZ",
                "Dr. Ogr. Uyesi Fatma Corut ERGIN",
                "Doc. Dr. Omer KORCAK",
                "Dr. Ogr. Uyesi Ali Haydar OZER",
                "Dr. Ogr. Uyesi Sanem Arslan"
        };
        //System.out.println("The size of the courseDtos is:"+courseDtos.size());
        //System.out.println("The size of the nteCourseDtos is:"+nteCourseDtos.size());
        //System.out.println("The size of the teCourseDtos is:"+teCourseDtos.size());
        //System.out.println("The size of the fteCourseDtos is:"+fteCourseDtos.size());
        //System.out.println("The size of the prerequisiteDtos is:"+prerequisiteDtos.size());
        //System.out.println("The size of the student names is:"+studentNames.size());
        for(int i=0;i<temp.length;i++){
            hireAdvisor(new Advisor(temp[i]));
        }

        for(CourseDto courseDto:courseDtos){
            courses.add(courseDto.convertToRealObject(prerequisiteDtos));
        }
        for(CourseDto courseDto:fteCourseDtos){
            fteCourses.add(courseDto.convertToRealObject(prerequisiteDtos));
        }
        for(CourseDto courseDto:nteCourseDtos){
            nteCourses.add(courseDto.convertToRealObject(prerequisiteDtos));
        }
        for(CourseDto courseDto:teCourseDtos){
            teCourses.add(courseDto.convertToRealObject(prerequisiteDtos));
        }
    }

    public void writeWholeSystem(Boolean beforeRegistration, String departmentOutput) throws IOException {
        String postfix = (beforeRegistration)? "before.json" : "after.json";
        //'before' means the student file before she finished her registration.
        //'after' means the student file after she finished her registration.
        //so we create two files for each student:
        //150120825before.json
        //150120825after.json
        //So, observing the results would become much easier.


        for(Student currStudent:students){
            String pathName = currStudent.getStudentID() + postfix;
            FileWriter writer = new FileWriter("iteration#1/data/student/"+pathName);
            Gson gson = new GsonBuilder().setPrettyPrinting().create();


            StudentDto studentDto = new StudentDto(currStudent);
            //With this single line of code Student object is converted to StudentDto object.
            //Now we are ready to write the student.
            gson.toJson(studentDto,writer);
            writer.close();
        }

        if(beforeRegistration==false){
            String pathName = "Department_Output_"+semester+".txt";
            FileWriter writer = new FileWriter("iteration#1/data/department/"+pathName);
            writer.write(departmentOutput);
            writer.close();
        }
    }

    public Student getStudentByID(String studentID){
        //Later we will get the student by reading the json file of that student.
        //For now, we just get the student by looking into the student array.
        for(Student searchStudent:students){
            if(searchStudent.getStudentID().equals(studentID)){
                return searchStudent;
            }
        }
        return null;
    }

    public Course getCourseByID(String courseID){
        for(Course course:courses){
            if(course.getCourseID().equals(courseID)){
                return course;
            }
        }
        return null;
    }

    public ArrayList<Course> getAllCourses(){
        return courses;
    }

    public ArrayList<Course> getAllTECourses(){
        return teCourses;
    }

    public ArrayList<Course> getAllFTECourses(){
        return fteCourses;
    }

    public ArrayList<Course> getAllNTECourses(){
        return nteCourses;
    }

    public ArrayList<Student> getAllStudents(){
        return students;
    }

    public Course getRandomFTECourse(){
        Random rand = new Random();
        Integer index = rand.nextInt(fteCourses.size());
        return fteCourses.get(index);
    }

    public Course getRandomTECourse(){
        Random rand = new Random();
        Integer index = rand.nextInt(teCourses.size());
        return teCourses.get(index);
    }

    public Course getRandomNTECourse(){
        Random rand = new Random();
        Integer index = rand.nextInt(nteCourses.size());
        return nteCourses.get(index);
    }

    public String getSemester() {
        return semester;
    }

    public Boolean getRandomPlay() {
        return randomPlay;
    }

    public Advisor getRandomAdvisor(){
        Random rand = new Random();
        Integer index = rand.nextInt(advisors.size());
        return advisors.get(index);
    }

    public String getRandomStudentName(){
        Random rand = new Random();
        Integer index = rand.nextInt(studentNames.size());
        return studentNames.get(index);
    }
}

File: Python/system/university.py
import json
import os
import random

from collections import OrderedDict
from model.course import Course

from model.advisor import Advisor
from system.prerequisiteTree import PrerequisiteTree

import logging
logging.basicConfig(filename='log.txt', encoding='utf-8', level=logging.DEBUG)

class University:
  def __init__(self):
      self.students = []
      self.courses = []
      self.nte_courses = []
      self.fte_courses = []
      self.te_courses = []
      self.advisors = []
      self.student_names = []
      self.student_transcripts = []
      self.semester = "fall"
      self.prerequisites = []

  def register_student(self,student):
    logging.info("Student {0} has been registered to university".format(student.get_student_id()))
    self.students.append(student)

  def hire_advisor(self,advisor):
    logging.info("Advisor {0} has been hired to university".format(advisor.advisor_name))
    self.advisors.append(advisor)

  def add_new_course(self,course):
    logging.info("Course {0} has been added to curriculum".format(course.course_id))
    self.courses.append(course)

  def read_all_input(self):
    input_file_path = os.path.join("Python","data", "input.json")
    f = open(input_file_path)
    input_file = json.loads(f.read())
    f.close()

    nte_courses=[]
    courses = []
    te_courses = []
    fte_courses = []
    prerequisites = []

    for key in input_file.keys():
        if(key == "semester"):
            self.semester = input_file[key]
        elif(key == "nonTechnicalElectiveCourses"):
            nte_courses = input_file[key]
        elif(key == "compulsoryCourses"):
            courses = input_file[key]
        elif(key == "technicalElectiveCourses"):
            te_courses = input_file[key]
        elif(key == "facultyTechnicalElectiveCourses"):
            fte_courses = input_file[key]
        elif(key == "prerequisites"):
            prerequisites = input_file[key]
        elif(key == "studentNames"):
            self.student_names =input_file[key]
    
    advisors = [
            "Prof. Dr. Haluk Rahmi TOCUOGLU",
            "Prof.Dr. Cigdem Eroglu ERDEM",
            "Doc. Dr. Ali Fuat ALKAYA",
            "Doc. Dr. Murat Can GANIZ",
            "Doc. Dr. Mujdat SOYTURK",
            "Doc. Dr. Borahan TUMER",
            "Doc. Dr. Mustafa AGAOGLU",
            "Dr. Ogr. Uyesi Mehmet BARAN",
            "Dr. Ogr. Uyesi Betul Demiroz BOZ",
            "Dr. Ogr. Uyesi Fatma Corut ERGIN",
            "Doc. Dr. Omer KORCAK",
            "Dr. Ogr. Uyesi Ali Haydar OZER",
            "Dr. Ogr. Uyesi Sanem Arslan"
    ]
    for i in range(len(advisors)):
        self.hire_advisor(Advisor(advisors[i]))
    
    for course in courses:
      self.courses.append(Course(course["courseCode"],
                                  course["courseName"],
                                  course["courseType"],
                                  course["credit"],
                                  course["semester"],
                                  self.find_prerequisites(prerequisites, course["courseCode"]),
                                  course["quota"]))
      
    for course in fte_courses:
      self.fte_courses.append(Course(course["courseCode"],
                                  course["courseName"],
                                  course["courseType"],
                                  course["credit"],
                                  course["semester"],
                                  self.find_prerequisites(prerequisites, course["courseCode"]),
                                  course["quota"]))
    
    for course in nte_courses:
      self.nte_courses.append(Course(course["courseCode"],
                                  course["courseName"],
                                  course["courseType"],
                                  course["credit"],
                                  course["semester"],
                                  self.find_prerequisites(prerequisites, course["courseCode"]),
                                  course["quota"]))
    for course in te_courses:
      self.te_courses.append(Course(course["courseCode"],
                                  course["courseName"],
                                  course["courseType"],
                                  course["credit"],
                                  course["semester"],
                                  self.find_prerequisites(prerequisites, course["courseCode"]),
                                  course["quota"]))
    
  def find_prerequisites(self, courses, key_course):
    for course in courses:
      if course["courseCode"] == key_course:
        return course["prerequisite"]
    return None

  def write_whole_system(self, before_registration, department_output):
    postfix = None
    if (before_registration):
      postfix = "before.json"
    else:
      postfix = "after.json"

    for cur_student in self.students:
      output_json_path = os.path.join("Python","data", "student", cur_student.get_student_id() + postfix)
      temp_dict = OrderedDict({})
      """self.student_name = student_name
        self.student_advisor = student_advisor
        self.student_id = StudentID(student_id)
        self.student_semester = student_semester
        self.added_courses = []
        self.student_transcript = StudentTranscript(self)
        self.weekly_schedule = [[None for x in range(5)] for y in range(8)]  # should be course array
        self.actions = []"""
      temp_dict["studentName"] = cur_student.student_name
      temp_dict["studentID"] = cur_student.get_student_id()
      temp_dict["studentSemester"] = cur_student.student_semester
      temp_dict["studentAdvisor"] = cur_student.student_advisor.advisor_name
      temp_dict["gpa"] = cur_student.get_cgpa()
      temp_dict["weeklySchedule"] = ({"Monday": [], "Tuesday": [], "Wednesday": [], "Thursday": [], "Friday": []})
      for hours in cur_student.weekly_schedule:
        day = 1
        for course in hours:
          if course != None:
            if day == 1:
              temp_dict["weeklySchedule"]["Monday"].append(course.course_id)
            elif day == 2:
              temp_dict["weeklySchedule"]["Tuesday"].append(course.course_id)
            elif day == 3:
              temp_dict["weeklySchedule"]["Wednesday"].append(course.course_id)
            elif day == 4:
              temp_dict["weeklySchedule"]["Thursday"].append(course.course_id)
            elif day == 5:
              temp_dict["weeklySchedule"]["Friday"].append(course.course_id)
          day += 1
      temp_dict["letterGrades"] = {}
      for grade in cur_student.student_transcript.get_letter_grades():
        temp_dict["letterGrades"][grade.course.course_id] = grade.letter_grade
      
      with open(output_json_path, 'w') as outfile:
        json.dump(temp_dict, outfile)

    if(before_registration == False):
      output_results_path = os.path.join("Python","data", "department", "department_output.txt")
      with open(output_results_path, 'w') as outfile:
        outfile.write(department_output)

  def get_student_by_id(self, student_id):
    for student in self.students:
      if(student.get_student_id() == student_id):
        return student
    return None

  def get_course_by_id(self, course_id):
    for course in self.courses:
      if(course.get_course_id() == course_id):
        return course
    return None

  def get_all_courses(self):
    return self.courses

  def get_all_te_courses(self):
    return self.te_courses

  def get_all_fte_courses(self):
    return self.fte_courses

  def get_all_nte_courses(self):
    return self.nte_courses

  def get_all_students(self):
    return self.students

  def get_random_fte_course(self):
    return self.fte_courses[random.randint(0, len(self.fte_courses)-1)]

  def get_random_te_course(self):
    return self.te_courses[random.randint(0, len(self.te_courses)-1)]

  def get_random_nte_course(self):
    return self.nte_courses[random.randint(0, len(self.nte_courses)-1)]

  def get_semester(self):
    return self.semester

  def get_random_advisor(self):
    return self.advisors[random.randint(0, len(self.advisors)-1)]

  def get_random_student_name(self):
    return self.student_names[random.randint(0, len(self.student_names)-1)]